lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,888 @@
+		let hash = self.hash;
+		let chaining_key = self.chaining_key;
+		let act_two = Act::from(act_two_builder);
+
+		let (responder_ephemeral_public_key, hash, chaining_key, temporary_key) = process_act_message(
+			&act_two,
+			&initiator_ephemeral_private_key,
+			chaining_key,
+			hash,
+		)?;
+
+		let mut act_three = EMPTY_ACT_THREE;
+
+		// start serializing act three
+		// 1. c = encryptWithAD(temp_k2, 1, h, s.pub.serializeCompressed())
+		chacha::encrypt(&temporary_key, 1, &hash, &initiator_static_public_key.serialize(), &mut act_three[1..50]);
+
+		// 2. h = SHA-256(h || c)
+		let hash = concat_then_sha256!(hash, act_three[1..50]);
+
+		// 3. se = ECDH(s.priv, re)
+		let ecdh = ecdh(&initiator_static_private_key, &responder_ephemeral_public_key);
+
+		// 4. ck, temp_k3 = HKDF(ck, se)
+		let (chaining_key, temporary_key) = hkdf5869rfc::derive(&chaining_key, &ecdh);
+
+		// 5. t = encryptWithAD(temp_k3, 0, h, zero)
+		chacha::encrypt(&temporary_key, 0, &hash, &[0; 0], &mut act_three[50..]);
+
+		// 6. sk, rk = HKDF(ck, zero)
+		let (sending_key, receiving_key) = hkdf5869rfc::derive(&chaining_key, &[0; 0]);
+
+		// 7. rn = 0, sn = 0
+		// - done by Conduit
+		let conduit = Conduit::new(sending_key, receiving_key, chaining_key);
+
+		// 8. Send m = 0 || c || t
+		act_three[0] = 0;
+		Ok((
+			Some(Act::Three(act_three)),
+			HandshakeState::Complete(Some((conduit, responder_static_public_key)))
+		))
+	}
+}
+
+impl IHandshakeState for ResponderAwaitingActThreeState {
+	// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-three (receiver)
+	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String> {
+		let mut act_three_builder = self.act_three_builder;
+		let bytes_read = act_three_builder.fill(input);
+
+		// In the event of a partial fill, stay in the same state and wait for more data
+		if !act_three_builder.is_finished() {
+			assert_eq!(bytes_read, input.len());
+			return Ok((
+				None,
+				HandshakeState::ResponderAwaitingActThree(Self {
+					hash: self.hash,
+					responder_ephemeral_private_key: self.responder_ephemeral_private_key,
+					chaining_key: self.chaining_key,
+					temporary_key: self.temporary_key,
+					act_three_builder
+				})
+			));
+		}
+
+		let hash = self.hash;
+		let temporary_key = self.temporary_key;
+		let responder_ephemeral_private_key = self.responder_ephemeral_private_key;
+		let chaining_key = self.chaining_key;
+
+		// 1. Read exactly 66 bytes from the network buffer
+		let act_three_bytes = Act::from(act_three_builder);
+		assert_eq!(act_three_bytes.len(), ACT_THREE_LENGTH);
+
+		// 2. Parse the read message (m) into v, c, and t
+		let version = act_three_bytes[0];
+		let tagged_encrypted_pubkey = &act_three_bytes[1..50];
+		let chacha_tag = &act_three_bytes[50..];
+
+		// 3. If v is an unrecognized handshake version, then the responder MUST abort the connection attempt.
+		if version != 0 {
```

Comment:
I would recommend all these above int literals be replaced by the constants defined in acts.rs. As for the version, I think a separate method `is_version_supported()`, would be better over just a näive `!= 0` check, which would also trivially allow supporting more versions in the future.

Reply:
To be honest, this seems like a bit of bikeshedding over 3 weeks into the review. I'll add my comments, but this is getting tedious to rebase/fixup with the number of trivial back-and-forths when this code is a big step forward and isn't a regression.

All of these values are directly referenceable from the RFC linked and it avoids littering the constants with ACT_THREE_CHACHA_START when they aren't used anywhere else.

As far as the version check. I also don't understand adding the overengineering. When multiple versions need to be supported, you can just write a test, it will fail, and you come to fix it. No need to add complexity and abstraction when the RFC only supports 0 and it is clearly referenced in the code and surrounding comments.

```
A version of 0 indicates that no change is necessary, while a non-zero version indicate that the client has deviated from the protocol originally specified within this document.
```

Reply:
I agree on the bikeshidding here. I don't think we'll update the `version` field soon from now.

Reply:
Fair point!